### ISSUE TITLE
pubsub: too many requests errors

### ISSUE THREAD

**Issue by horgh** [googleapis/google-cloud-go#10610]
**Client**

PubSub

**Environment**

Ubuntu 20.04 on GCE VM

**Go Environment**

```
$ go version
go version go1.22.5 linux/amd64
```

```
$ go env
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/home/wstorey_maxmind_com/.cache/go-build'
GOENV='/home/wstorey_maxmind_com/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/home/wstorey_maxmind_com/go/pkg/mod'
GONOPROXY=''
GONOSUMDB='github.maxmind.com/*'
GOOS='linux'
GOPATH='/home/wstorey_maxmind_com/go'
GOPRIVATE=''
GOPROXY='https://artifactory.maxmind.com/artifactory/api/go/go/'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/local/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.22.5'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='1'
GOMOD='/home/wstorey_maxmind_com/mm_website/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build3119879897=/tmp/go-build -gno-record-gcc-switches'
```

**Code**

I don't have a reproducer currently. It's happening deep in some of our daemons that have been using Pub/Sub for some time.

**Expected behavior**

No 429 errors.

**Actual behavior**

We are occasionally receiving this error:

```
rpc error: code = Unauthenticated desc = transport: per-RPC creds failed due to error: compute: Received 429 `"Too many requests."`
```

**Additional context**

We receive these errors occasionally. Since 00:00:00 UTC today (it's 18:53 as I type) we've received 173 of them.

This started after we upgraded from cloud.google.com/go/pubsub v1.39.0 to v1.40.0. When we updated that, a few other cloud.google.com packages were updated, but I'm not sure if they're relevant:

* iam from v1.1.8 to v1.1.10
* compute from v1.27.0 to v1.27.2
* compute/metadata from v0.3.0 to v0.5.0
* auth from v0.5.2 to v0.7.1

Given the similarity to the error in #10544, I am wondering if it's the auth package update.

I'm still trying to get more details and reproduce it, but I wonder if you have any thoughts.

Thank you!

**Comment by horgh** [comment]
This is an error coming from `Get()` on the result of a topic `Publish()` call.

**Comment by hongalex** [comment]
It seems like there was a fix, https://github.com/googleapis/google-cloud-go/pull/10551, that was released as `auth v0.72.0` that might have fixed this. Could you upgrade and see if that changes?

**Comment by horgh** [comment]
We're actually using auth v0.7.2 already. The v0.7.1 I mentioned was in the deploy when we first began seeing this error, but it's still happening with v0.7.2 deployed too.

Of the others I mentioned, we're using iam v1.1.11, compute v1.27.3, and compute/metadata v0.5.0.

I think the error is a bit different to the one in #10551 as it's a 429 rather than a connection error or 403.

**Comment by horgh** [comment]
It did not stabilize. Between midnight Pacific and 5am Pacific (long after we last deployed and before any deploy today) we had it happen 39 times. It actually seems to be becoming more frequent since approximately July 27 (first began July 19 after the update I mentioned). Our usage patterns have not changed AFAIK.

Would you have any advice about trying to debug it?

**Comment by horgh** [comment]
Also the publishers in question are long running daemons. The one that happened to be logging over the night last night for example would not have restarted or anything.

**Comment by codyoss** [comment]
> Would you have any advice about trying to debug it?

It may be worth reaching out to customer support as they can get a deeper understanding of what is happening in your account. Only thing I can tell from the error is there is likely an influx of metadata service calls.

**Comment by horgh** [comment]
Thank you, I'll ask them!

**Comment by horgh** [comment]
I was also thinking of looking into where the metadata calls are happening when publishing. I wonder if it is happening more often than necessary for some reason. I wouldn't expect to have to hit it often enough to be a problem!

**Comment by horgh** [comment]
I haven't had success replicating any large number of requests to the metadata server. I currently believe the most likely request is this one: `http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token?scopes=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform%2Chttps%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpubsub`. But that is just a theory based on it being the only one I see occurring during publishing (I think at stream open?). We use the VM service account.

We opened a ticket July 31. They say they see no 429 responses from the metadata server from the instances in question. They also said:

> You also mentioned your devs are confident the error comes from the metadata server, could you please elaborate on why they believe this. I'm trying to connect dots, as logs from the metadata server show no out of the ordinary activity.

Given the error says `compute: Received: ` which appears to only come from the metadata package, I am fairly sure that as you suggest, it's requests to the metadata server. It would be helpful if I could show definitively that is the case though.

Would adding the attempted request URL to the error message be an option? I suppose I could hack up the package myself, but I'm wondering if it could be helpful generally for debugging issues like this.

Do you think that stream opens are the only time I should expect metadata requests to happen in a daemon where the only use of Pub/Sub is to publish?

Interestingly, we have now gone over 29 hours without the error occurring. This is the longest stretch we've gone since it began. I'll attach a graph.

![Screenshot 2024-08-02 08 17 08](https://github.com/user-attachments/assets/48a08f8a-8a64-4605-8a38-0241fbf9820e)

All of these errors save 3 were from the Pub/Sub publish calls. Of the remaining 3, 2 were attempted requests to the storage API and 1 was an attempted request to the compute API - all Go code showing the `compute: Received 429` error messages, so presumably also metadata requests failing due to metadata server issues. Since the remaining 989 come from attempted Pub/Sub publishes, I am still most suspicious of that.

**Comment by horgh** [comment]
Thanks! The case is 52663090.

Yes, I surmised the 4 count given the mention of 4 streams in the docs. It's odd we'd hit a 429 frequently in the Pub/Sub code if it's just 4x an hour per daemon.

In one of the instances that we've observed it, we have 1 daemon with 1 client that uses Pub/Sub. There is one other daemon using different GCP API(s) on there I believe, but not much else. Several daemons on there are using cloud tracing. Possibly there are some other APIs in use that I don't immediately see though.

Do you think there's any way to enable tracing that would help?

**Comment by horgh** [comment]
Well we have seen no further cases of this error since the ones in the last graph I posted. Apparently no 429s were visible according to the support case I mentioned.

I'm not sure what to make of this stopping other than maybe something was fixed on the GCP side that I don't have visibility into.

I suppose we can close this given it seems to indeed not be a bug in the package. If you were able to find any further info about what was going on, I'd definitely be curious!

Thank you for your help!
